### DISCUSSION TITLE
Block width in martensite (assumed dimension of grain perpendicular to selected plane)

### DISCUSSION THREAD

**Discussion by nyyssont** [mtex-toolbox/mtex#522] [Feature Requests]
Dear MTEX community,

as you know, MTEX reconstructs grains from EBSD orientation maps. In the case of martensite formed in low-carbon steels, these grains are in literature often called "blocks", with each block containing two types of martensitic variants with a low mutual misorientation angle. The variants in a block form side by side into a stack of thin laths so that on the image plane, the block typically has a clearly distinguishable long axis and a short axis. "Block width" is the width of the stack of laths in the block, NOT the length of the short axis of the block on the image plane. The stack of laths that makes up the block forms approximately on the particular {111} plane in austenite that is parallel to the (011) plane in martensite. Please see the attached block width schematic to help visualize the situation.

The new prior austenite reconstruction features allow us to easily determine the specific {111} type plane in austenite parallel to the (011) plane in martensite. The trace and normal of this specific {111} plane can be plotted for visual verification. The attached figure shows that the long axis of each block indeed appears to be parallel to a particular {111} plane.

I'd like to characterize block width. This would entail getting some sort of representative value of the dimension of each block perpendicular to the trace of the particular {111} plane. A value for block width could then be generated by assuming that the block width divided by the dimension on the image plane equals cosine of the angle between the particular {111} plane and the image plane.

I hope the above explanation and the attached images made sense to someone. 

To summarize:
1) I'm looking for block width and
2) I would appreciate any suggestion on how to easily and efficiently get the dimensions of each block perpendicular to the {111} trace on the image plane.

I've attached the code used to generate the image with the plane traces (block_width.zip).

[block_width.zip](https://github.com/mtex-toolbox/mtex/files/5378697/block_width.zip)
![block_width_schematic](https://user-images.githubusercontent.com/10060198/96000336-10a26380-0e37-11eb-8203-a7b18551411b.png)
![111_planes_in_PAG](https://user-images.githubusercontent.com/10060198/96000473-392a5d80-0e37-11eb-8c0b-5de168419d37.jpg)

**Comment by kilir** [comment]
Hi,
What would you consides the width of a somewhat irregular, non-rectangular grain, eg the orange, lense shaped one in the left part og your image? Would the projection ( the "shadow") of the grain parallel to the trace of {111} be ok?
Cheers Rüdiger

**Reply by nyyssont** [reply]
That's a good question. In the case of the orange-colored grain, I could imagine that the stack of laths making up the block lies almost flat on the image plane, so it's a bit hard to imagine how to measure the block width (the thickness of the stack).

But I would probably want to measure the width similarly for every block (perpendicular to plane) and then later on make up some kind of quality metric that could be used to disqualify cases where the block lies almost flat on the image plane (or to put another way, the particular {111} plane lies almost flat on the image plane).

**Reply by kilir** [reply]
Ok, but let's consider the yellow grains in the center with almost vertical long axes. They have embayments and have thinner and thicker parts (normal to the trace of {111}). Which would be the measure to consider? The largest width of the grain?
Cheers,
Rüdiger

**Reply by nyyssont** [reply]
Yeah, that's another good one. I imagine the largest width normal to trace would be easiest to get? I would perhaps like some sort of distribution of width normal to trace for each grain, mean value etc. I can definitely see a problem with, for example, the yellow grain that wraps around a smaller yellow grain. How to characterize that one and others like it?

Anyway, I guess largest width normal to trace would be a good starting point for this type of analysis. Any other martensite people around with opinions?

-Tuomo

**Comment by nyyssont** [comment]
It looks like the function caliper described on this page could do the trick:
https://mtex-toolbox.github.io/ProjectionBasedParameters.html

However, caliper doesn't seem to exist in MTEX 5.5.beta.3. There is a function called "calliper" but it doesn't work quite as described on the page.

I suppose this functionality is still under development?

**Reply by nyyssont** [reply]
I see there is a function "caliper" in the development version. Will test if it could produce something nice...

**Comment by nyyssont** [comment]
Hello,

attached is an attempt to calculate block width by directing caliper against the azimuthal angle of the correct {111}austenite plane normal and then multiplying said caliper value by the sine of the inclination angle of said plane normal. This uses the "caliper" in the development version of MTEX.

The plotted image shows the grains colored with the calculated block width value, as well as the normal and trace of the {111} plane. The caliper on the image plane is shown as well. As expected, a grain with a large caliper value but with the correct {111} plane close to parallel with the image plane will get a smaller block width value.

I would appreciate comments and suggestions on how to improve this or whether you think the script is actually achieving what I want.

Updated wishlist:
1. "Shortest" caliper value in specified omega direction
2. Option in "quiver" to disable 'autoScale' and just present the given vector in its full length.

![block_width_calculated](https://user-images.githubusercontent.com/10060198/96846848-b87cea00-1452-11eb-88ea-4fd3da6322a7.jpg)
![block_width_histogram](https://user-images.githubusercontent.com/10060198/96846862-be72cb00-1452-11eb-9f65-7c9cbd0c9d73.jpg)
[block_width.zip](https://github.com/mtex-toolbox/mtex/files/5421433/block_width.zip)

**Reply by ralfHielscher** [reply]
Hi Tuomo,

1. I do not really get it - what is the shortest caliper in a certain direction. You can have the caliper or projection length in a certain projection - but this will not be the shortest for most angles. 
```matlab
l = caliper(grains,omega)
```
2. You can turn of autoScaling with the option ```'noScaling'```

Ralf.

**Reply by ralfHielscher** [reply]
PS: Any chance you include this into the documentation - as a new page :smile:

**Reply by nyyssont** [reply]
Not a problem, I can write this out when I'm done with the scripts.

Regarding item 1:
I'd be interested to get a range of values for the width of the grain in the direction of interest. Kind of like doing many line intercepts, with the hope that there would be some more elegant way to obtain these values. The attached image hopefully visualizes what I'm after.
![width_range](https://user-images.githubusercontent.com/10060198/96863242-d81e0d80-1466-11eb-8376-a9ae9aab1b50.jpg)

**Reply by ralfHielscher** [reply]
Ok, I see. I'm not aware of an elegant algorithm for that. But looks interesting.

Ralf.

**Reply by kilir** [reply]
Couldn't you project all points of the boundary onto the trace of the plane normal. So one would get a histogram - bimodal in the case of a lath shaped grain - and from there one could decide which distance to take. If it's not a lath with it's long axis normal to the projection direction, it will be a unimodal, broad distribution.
Rüdiger

**Reply by nyyssont** [reply]
Thanks for the suggestion Rudiger, that sounds like exactly what I want. I will look into it. I wonder if you have any hint on how to accomplish that, script-wise :D

**Reply by kilir** [reply]
Can you check if this is reasonable? 
```
function p = projectPoints2Caliper(grains,v)

% output cell
if length(grains) ~= length(v), error('number of input must be identical'); end
% v can come from caliper

V = grains.V;
poly = grains.poly;
c= grains.centroid;

for i=1:length(grains)
    
% get vertices
Vg = V(poly{i},:);

% center vertices

Vg = [Vg(:,1)-c(i,1) Vg(:,2)-c(i,2)]

% r = rotation.byAxisAngle(zvector,v(i).rho)
%omega = -v(i).rho;
%r = [cos(omega) -sin(omega); sin(omega) cos(omega)];
%Vg = r*Vg'; % rotate all points by - angle
% project (count numbers on x or y?)
%p{i}=Vg(y,:);

p{i} = cos(-v(i).rho)*Vg(:,1) -sin(-v(i).rho)*Vg(:,1);

end
end
```

Example:
```
v = caliper(grains,'shortest');
p = projectPoints2Caliper(grains,v)
```
Will give you a cell for each grain with a bunch of numbers.

Cheers,
Rüdiger

**Reply by nyyssont** [reply]
This seems to work quite nicely. Taking the mean of the absolute values and multiplying by two seems to consistently give a bit lower number than caliper, as expected. The Figure below shows this value in the measurement direction. The potential ability to identify lath shape by histogram bimodality is a nice additional bonus.

Calculating the average block width gives a value of about 1,5 um. This corresponds to the values calculated in earlier work for a (somewhat) similar steel by Morito et al [1]. In their study, I believe the plane normals and block width were determined by manual work and calculations. It's nice to reach a similar result automatically.

I'm really happy now! Thanks!

I'll refine this a bit and will attempt to write some documentation.

[1] S. Morito et al. / Materials Science and Engineering A 438–440 (2006) 237–240
![block_width_calculated_new](https://user-images.githubusercontent.com/10060198/96917626-6ca95f80-14a9-11eb-961f-380799898480.jpg)

**Reply by nyyssont** [reply]
Actually, looks like the function gives too large values to some grains (such as the mustard-colored grain above). Changing the function to this fixes the issue:
```matlab
function p = projectPoints2Caliper(grains,v)

% output cell
if length(grains) ~= length(v), error('number of input must be identical'); end
% v can come from caliper

V = grains.V;
poly = grains.poly;
c= grains.centroid;

for i=1:length(grains)
    
    % get vertices
    Vg = V(poly{i},:);
    
    % center vertices
    
    Vg = [Vg(:,1)-c(i,1) Vg(:,2)-c(i,2)];
    a = v(i).y/v(i).x;
    b = -1;

    p{i} = (a*Vg(:,2) - b*Vg(:,1))./sqrt(a^2+b^2);

    % r = rotation.byAxisAngle(zvector,v(i).rho)
    %omega = -v(i).rho;
    %r = [cos(omega) -sin(omega); sin(omega) cos(omega)];
    %Vg = r*Vg'; % rotate all points by - angle
    % project (count numbers on x or y?)
    %p{i}=Vg(y,:);
    
%     p{i} = cos(-v(i).rho)*Vg(:,1) -sin(-v(i).rho)*Vg(:,1);
    
end
end
```
The average calculated block size now becomes 1.28 um, which is still about as expected.

![projectedpoints_latest](https://user-images.githubusercontent.com/10060198/96995092-f00c9480-152d-11eb-8daf-b0e6488cbfa8.jpg)

**Reply by kilir** [reply]
Hi,
yes, you are right, there was a tyo:

```
p{i} = cos(-v(i).rho)*Vg(:,1) -sin(-v(i).rho)*Vg(:,1);
```
should be
```
p{i} = cos(-v(i).rho)*Vg(:,1) -sin(-v(i).rho)*Vg(:,2);
```


Here's a simple test case:
```
mtexdata testgrains
%%
v = caliper(grains(8),'shortest')
p = projectPoints2Caliper(grains(8),v)

plot(grains(8))
hold on
quiver(grains(8),v,'autoScaleFactor',0.5)
hold off

v.norm
figure
histogram(p{1},100)
```

Cheers,
Rüdiger

**Comment by nyyssont** [comment]
OK, I think I have it working the way I want now. I submitted a pull request with documentation and a function projectPoints2Vector. Thanks for your help.

Comments, critique and suggestions on implementation and functionality of this approach to block width characterization are still welcome!

**Reply by ArianeNeves** [reply]
Hi Tuomo!
I have been trying to use your code (https://github.com/mtex-toolbox/mtex/pull/544).

Is there a way to find the matching (x, y) according to habit plane of martensite? I see that you use the vertices.

![Largura do bloco](https://user-images.githubusercontent.com/81368454/180790054-0e2cb289-112d-42b0-b364-6ca0dcd32f1f.jpg)

In my case, 5 lines are good. 

I found this function https://www.mathworks.com/matlabcentral/fileexchange/93470-intersections-of-multiple-2d-lines-or-line-segments by Matt J (2022). But, it works only for a regular shapes (x max and y max fixed).
In this way, it is possible to use the spatialProfile function to measure the apparent width.
